[GUI][Refactoring] Constrain direct wallet access from the GUI in a single place

random-zebra opened the pull request:
First step for the multiwallet implementation.
We have several places in the GUI code, where we freely access `pwalletMain`.
Such global should be used exclusively in `pivx-qt.cpp` to initialize the internal pointer of `WalletModel`. All other calls to the wallet from the GUI should pass through the model.

Bonus: In the process, kill multisend with fire.

furszy commented on src/qt/walletmodel.cpp:
```diff
@@ -322,6 +336,26 @@ void WalletModel::setWalletCustomFee(bool fUseCustomFee, const CAmount& nFee)
+    LOCK(wallet->cs_wallet);
+    if (wallet->nStakeSplitThreshold != nStakeSplitThreshold) {
+        wallet->nStakeSplitThreshold = nStakeSplitThreshold;
+        CWalletDB(wallet->GetDBHandle()).WriteStakeSplitThreshold(nStakeSplitThreshold);
```
now that we are here, would be good to encapsulate the `nStakeSplitThreshold` field access and the wallet db write access.

random-zebra replied:
yep, done.

furszy commented on src/qt/walletmodel.cpp:
```diff
@@ -54,6 +54,20 @@ WalletModel::~WalletModel()
+    optionsModel->refreshDataView();
+}
+
+void WalletModel::resetOptionsToDefault(QSettings& settings)
```
Wouldn't be better to call it `resetWalletOptions()` ? So we know that the options are only for the wallet.

random-zebra replied:
done.

furszy commented on src/qt/optionsmodel.cpp:
```diff
@@ -562,6 +525,22 @@ bool OptionsModel::isRestartRequired()
+    QSettings settings;
+    double oldVal = getSST();
+    if (oldVal != val) {
+        settings.setValue("SSTVal", val);
```
we shouldn't be duplicating the stake split threshold storage, and only store it in the walletdb.

- I know that this is not from this PR and can be tackled later -

random-zebra replied:
done.

furszy commented on src/qt/walletmodel.cpp:
```diff
@@ -54,6 +54,20 @@ WalletModel::~WalletModel()
+{
+    // update optionsModel with wallet options
+    optionsModel->setSST(static_cast<double>(getWalletStakeSplitThreshold() / COIN));
+    optionsModel->refreshDataView();
```
`refreshDataView` will refresh the data for every element in the settings. Better to only update the view for the SST row.

random-zebra replied:
removed the refresh completely, and added the signal update.

furszy commented on src/qt/pivx/settings/settingswalletoptionswidget.cpp:
```diff
@@ -65,6 +65,11 @@ void SettingsWalletOptionsWidget::onResetClicked()
 
+void SettingsWalletOptionsWidget::showEvent(QShowEvent *event)
+{
+    walletModel->loadWalletOptions();
```
I'm a NACK for this change, commit 8ef86697f84627e48e22aeee8089b28d5b025c03.
It's the same change that was introduced in 2239, check my comment there https://github.com/PIVX-Project/PIVX/pull/2239#issuecomment-797758921

random-zebra replied:
Implemented signal-update and did some more cleanup.

furszy replied:
awesome 👌 👌

furszy reviewed (commented):
Code reviewed up until 3ffd73038f1ce5f339afa2f28057bfb3d871ea94 , left few comments.

Other than that, looking good ☕ . One step closer to the multi-process goal :).

furszy commented on src/qt/pivx/settings/settingswalletoptionswidget.cpp:
```diff
@@ -79,11 +76,27 @@ void SettingsWalletOptionsWidget::setMapper(QDataWidgetMapper *mapper){
 
+void SettingsWalletOptionsWidget::setWalletModel(WalletModel* model)
+{
+    PWidget::setWalletModel(model);
```
Can override `loadWalletModel` to avoid doing this parent call.

Essentially, `PWidget::setWalletModel` calls internally to `loadWalletModel` which is the function that has to be overrode by its child if any setup is needed after the walletModel set (same workflow applies to the `clientModel` set).

furszy commented:
Checked c3f6a7db3f092ef7a594f0fb2d5659764ff3b8c2 and is looking quite good ☕ .

Have made some changes to it so we can get rid off the friend class declaration https://github.com/furszy/PIVX/commit/71bec543c770bc1c95aabb3cba486a71df3e66a7. Better to not allow private/protected members access/modifications from outside of the object and get a better encapsulation and class responsibilities division :).
Feel free to squash it inside that same commit, no need to cherry-pick it.

random-zebra commented:
Squashed @furszy's patch, with a little edit, inside 9638333c99e501dc786f961d659cceb39a6afc06.
Also rebased on master, fixing minor conflict after #2253 merge.

furszy reviewed (approved):
nice work! ☕☕ 
Code review ACK 9d7e5451dad14b9bee8e419322bdafeb76fd7be3

Will try it a bit now.

furszy commented:
Needs rebase

random-zebra commented:
Rebased.

furszy reviewed (approved):
re-utACK 012de60 after rebase.

Fuzzbawls reviewed (approved):
ACK 012de60777fe5cf6c639e7780e3ad3338db3cd47
